### DISCUSSION TITLE
Using native image with javax.mail and logback variables

### DISCUSSION THREAD

**Discussion by adamkobor** [micronaut-projects/micronaut-core#3937] [Q&A] [answered]
Hey there, 
I'm kind of stuck with two separate issues regarding GraalVM and native images. I have a pet project, called Kuvasz, you can find the public repo here: https://github.com/kuvasz-uptime/kuvasz

**The first problem is about logback and environment variables**. As you can see [here](https://github.com/kuvasz-uptime/kuvasz/blob/main/src/main/resources/logback.xml#L2), I'm using a variable like this in my `logback.xml`: 
```xml
<variable name="communicationLogLevel" value="${HTTP_COMMUNICATION_LOG_LEVEL:-INFO}"/>
```

The default value can be overridden in a "regular" docker image, but it doesn't work inside a native image (there is no exception, or anything suspicious, it just always uses the default value).

**The second problem is about `javax.mail`**. I'm using simple-java-mailer to send emails through a pre-configured SMTP host. There were issues about build-time class initialization and missing resources, but I already solved them (or at least I think so, because there are no build time errors). However, when I try to run the native image with a valid SMTP configuration, I get this at the bottom of the stacktrace:

```
kuvasz_1  | Caused by: javax.mail.NoSuchProviderException: smtp
kuvasz_1  |     at javax.mail.Session.getService(Session.java:842)
kuvasz_1  |     at javax.mail.Session.getTransport(Session.java:772)
kuvasz_1  |     at javax.mail.Session.getTransport(Session.java:713)
kuvasz_1  |     at javax.mail.Session.getTransport(Session.java:693)
kuvasz_1  |     at javax.mail.Session.getTransport(Session.java:673)
kuvasz_1  |     at org.simplejavamail.mailer.internal.util.TransportRunner.runOnSessionTransport(TransportRunner.java:74)
kuvasz_1  |     at org.simplejavamail.mailer.internal.util.TransportRunner.connect(TransportRunner.java:60)
kuvasz_1  |     at org.simplejavamail.mailer.internal.TestConnectionClosure.executeClosure(TestConnectionClosure.java:54)
```

(The mentioned part of the code in `Session.java` is using reflection heavily, and trying to instantiate a class with a `ClassLoader`)

In both cases I've tried to run the native image agent, but I can't get closer to the solution. (I've also attached the output of the agent.)

[native-image-agent-output-kuvasz.zip](https://github.com/micronaut-projects/micronaut-core/files/5103224/native-image-agent-output-kuvasz.zip)

Any help appreciated 🙏 

Micronaut version: 2.0.1
GraalVM version: graalvm-ce:20.2.0-java11

**Comment by fernando-valdez** [comment, marked as answer]
Hello @adamkobor, regarding the second problem:
The error `javax.mail.NoSuchProviderException: smtp`  could mean that some JDK classes are not included in the native image by default, and that is intended. 
We have caller-based filters you can use with the native-image-agent to get the proper configuration files. If you want to know more about this see `https://github.com/oracle/graal/blob/master/substratevm/CONFIGURE.md#caller-based-filters`
I would recommend you to initially use this filter:
```
{
    "rules": [
      {"includeClasses": "javax.mail.**"}
    ]
}
```
You will need to save it to a file, and then use native-image-agent like this:
java -agentlib:native-image-agent=config-output-dir=$CONFIG_DIR,caller-filter-file=$FILTER_FILE_PATH
Note the addition of caller-filter-file= for the filter path.

**Reply by adamkobor** [reply]
Thanks, will check it 🙏

**Reply by adamkobor** [reply]
Hey @fernando-valdez, it worked like a charm 🤗  Actually, I've ended up with the following `reflect-config.json` to make it work: https://github.com/kuvasz-uptime/kuvasz/blob/main/src/main/resources/META-INF/native-image/com.kuvaszuptime/kuvasz-application/reflect-config.json

**Reply by fernando-valdez** [reply]
Awesome! Thanks for reporting back!

**Reply by darren7690** [reply]
Hey there,
 I also had a problem about javax.mail. I'm using simple-java-mailer to send emails through a pre-configured SMTP host. There was issue about encoding for Chinese charactor: it works well in java jar; However, when I try to run the native image ,the Chinese charactors in email content  shows garbage code as ‘？’

Any help appreciated 🙏

**Comment by ilopmar** [comment]
@adamkobor regarding the logback issue. Have you tried asking/opening an issue in their tracker? Maybe it's something logback doesn't support yet, but they are not aware.

With the email thing, as @fernando-valdez recommended, use the native image tracing agent. More info here: https://medium.com/graalvm/introducing-the-tracing-agent-simplifying-graalvm-native-image-configuration-c3b56c486271

**Reply by adamkobor** [reply]
Thank you Iván, I'll check it tomorrow, and will also take a look at logback's GitHub.

**Reply by adamkobor** [reply]
A quick update: I've also tried to set this variable programmatically through logback's context, but I didn't succeed, so I'll head to the logback guys with the issue. Thanks for your help anyway!
